### PR TITLE
[MLIR] Add pattern to fold insert_slice of extract_slice

### PR DESCRIPTION
**Pull request by pzread**
Fold the `tensor.insert_slice` of `tensor.extract_slice` into `tensor_extract_slice` when the `insert_slice` simply expand some unit dims dropped by the `extract_slice`.

For example:

```mlir
%extracted_slice = tensor.extract_slice %in[0, 0] [1, 8] [1, 1] : tensor<2x8xf32> to tensor<8xf32>
%inserted_slice = tensor.insert_slice %extracted_slice into %dest[0, 0] [1, 8] [1, 1] : tensor<8xf32> into tensor<1x8xf32>
```

can be folded into:

``` mlir
%extracted_slice = tensor.extract_slice %in[0, 0] [1, 8] [1, 1] : tensor<2x8xf32> to tensor<1x8xf32>
```

### CODE REVIEW

**Review by matthias-springer — COMMENTED** [review]
I think this pattern would fit best next to `DropRedundantInsertSliceRankExpansion` (maybe rename `populateDropRedundantInsertSliceRankExpansionPatterns`).

**Comment by pzread** [comment]
> I think this pattern would fit best next to `DropRedundantInsertSliceRankExpansion` (maybe rename `populateDropRedundantInsertSliceRankExpansionPatterns`).

Done.

File: mlir/lib/Dialect/Tensor/Utils/Utils.cpp [hunk 1]
```diff
@@ -142,11 +142,15 @@ mlir::tensor::getUnPackInverseSrcPerm(UnPackOp unpackOp,
...
+  RankedTensorType resultType = op.getDestType();
   // Source dims and destination dims (apart from dropped dims) must have the
   // same size.
-  for (int64_t resultDim = 0; resultDim < op.getDestType().getRank();
-       ++resultDim) {
+  for (int64_t resultDim = 0; resultDim < resultType.getRank(); ++resultDim) {
     if (droppedDims.test(resultDim)) {
+      // InsertSlice may expand unit dimensions that result from inserting a
+      // size-1 slice into a non-size-1 result dimension.
+      if (resultType.getDimSize(resultDim) != 1)
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Tensor/Utils/Utils.cpp:152]: Why is this needed? If a dimension is "dropped", it must be a unit dimension.

Comment (pzread) [reply mlir/lib/Dialect/Tensor/Utils/Utils.cpp:152]: This is to check if the dropped unit dim is inserted into non-unit dim. For example:

```mlir
tensor.insert_slice %x into %y[0, 0, 0] [1, 8, 8] [1, 1, 1] : tensor<8x8xf32> into tensor<2x8x8xf32>
```

in this case, the first dim of `1x8x8xf32` is considered as the dropped unit dim, but it is inserted into `2x8x8xf32`. I think in this case we don't consider it as a cast-like op as the source `(1)x8x8xf32` != the destination `2x8x8xf32`

File: mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp [hunk 2]
```diff
@@ -134,6 +134,86 @@ struct DropRedundantInsertSliceRankExpansion
...
+
+    // Check if the insert_slice op purely expands ranks (add unit dims).
+    if (!isCastLikeInsertSliceOp(insertSliceOp))
+      return failure();
+
+    llvm::SmallBitVector extractDroppedDims = extractSliceOp.getDroppedDims();
+    llvm::SmallBitVector insertExpandedDims = insertSliceOp.getDroppedDims();
+    // Can't fold if the insert_slice op expands to more dims.
+    if (extractDroppedDims.size() < insertExpandedDims.size())
+      return failure();
```
Comment (hanhanW) [inline mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp:177]: I think it would be good if we put the failure reasons to `rewriter.notifyMatchFailure()`. I found it useful with `-debug`. Otherwise, we'll need to come back to the code and see why it's not matched.

Comment (pzread) [reply mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp]: Done.

**Review by hanhanW — APPROVED** [review]
LGTM, thanks

**Review by matthias-springer — APPROVED** [review]
